src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -17720,6 +17720,1151 @@ public void SimpleWhere()
                 );
         }
 
+        [Fact]
+        public void MaybeNullWhenTrue_Simple()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Runtime.CompilerServices;
+public class C
+{
+    public void Main(string s)
+    {
+        if (IsNull(s))
+        {
+            s.ToString(); // warn
+        }
+        else
+        {
+            s.ToString(); // ok
+        }
+
+        s.ToString(); // ok
+    }
+    public static bool IsNull([MaybeNullWhen(true)] string s) => throw null!;
+}
+", MaybeNullWhenAttributeDefinition }, options: WithNonNullTypesTrue());
+
+            c.VerifyDiagnostics(
+                // (9,13): warning CS8602: Dereference of a possibly null reference.
+                //             s.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(9, 13)
+                );
+        }
+
+        [Fact]
+        public void MaybeNullWhenTrue_Indexer()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Runtime.CompilerServices;
+public class C
+{
+    public void Main(string s)
+    {
+        _ = this[s]
+            ? s.ToString() // 1
+            : s.ToString();
+    }
+    public bool this[[MaybeNullWhen(true)] string s] => throw null!;
+}
+", MaybeNullWhenAttributeDefinition }, options: WithNonNullTypesTrue());
+
+            c.VerifyDiagnostics(
+                // (8,15): warning CS8602: Dereference of a possibly null reference.
+                //             ? s.ToString() // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(8, 15)
+                );
+        }
+
+        [Fact]
+        public void MaybeNullWhenTrue_OutParameter_MiscTypes()
+        {
+            // Warn on misused nullability attributes (F3, F4 and F5)? https://github.com/dotnet/roslyn/issues/36073
+            var source =
+@"using System.Runtime.CompilerServices;
+class Program
+{
+    static bool F1<T>(T t, [MaybeNullWhen(true)]out T t2) => throw null!;
+    static bool F2<T>(T t, [MaybeNullWhen(true)]out T t2) where T : class => throw null!;
+    static bool F3<T>(T t, [MaybeNullWhen(true)]out T? t2) where T : class => throw null!;
+    static bool F4<T>(T t, [MaybeNullWhen(true)]out T t2) where T : struct => throw null!;
+    static bool F5<T>(T? t, [MaybeNullWhen(true)]out T? t2) where T : struct => throw null!;
+    static void M1<T>(T t1)
+    {
+        _ = F1(t1, out var s2)
+            ? s2.ToString() // 1
+            : s2.ToString(); // 2
```

Comment:
>s2.ToString(); // 2 [](start = 14, length = 19)

I don't understand why there would be a warning here?

Reply:
Imagine that `M1` is invoked with `T = string?`. Then we're looking at `F1<string?>(t1, out string? t2)` which yields a maybe-null `t2` in both branches.

---
In reply to: [291751778](https://github.com/dotnet/roslyn/pull/36172#discussion_r291751778) [](ancestors = 291751778)